### PR TITLE
contrib/utilities/indent: Reorganize script

### LINKED ISSUE #6760: clang-format and recompilation
I don't have time to look into it but it looks like running ``indent`` causes the whole project to be marked dirty which forces a recompilation. Can we get this fixed?

### PR DESCRIPTION
**Pull request by tamiko**
This commit

 - reorganizes the core indenting logic into a indent_common.sh file

 - cleans up find and indenting logic to make everything more readable

 - prepares for more specialized indent versions (see #6673)

Supersedes #6761
Closes #6760
In preparation for #6673
In preparation for #6698

### CODE REVIEW

**Comment by tamiko** [comment]
@dealii/dealii 

This is now ready for review. The most important change is that everything is now aggressively encapsulated in a `process()` function:
```
process()
{
  find ${1} -regextype egrep -regex "${2}" -print0 |
    xargs -0 -n 1 -P 10 -I {} bash -c "${3} {}"
}
```
A typical call to that function is `process "include source" ".*\.(cc|h|cu|cuh)" format_file` and should be pretty self-explanatory. With that we can implement @drwells indent-branch variant in about 10 lines:
```
process_changed()
{
  LAST_MERGE_COMMIT="$(git log --format="%H" --merges --max-count=1 master)"

  ( git ls-files -z --others --exclude-standard -- ${1};
    git diff -z --name-only $LAST_MERGE_COMMIT -- ${1} )|
      sort -zu |
      grep -zE "${2}" |
      xargs --no-run-if-empty -0 -n 1 -P 10 -I {} bash -c "${3} {}"
}
```

File: contrib/utilities/indent [hunk 1]
```diff
@@ -23,129 +23,39 @@
...
-  tr -d '\015' <"$f" >"$f.tmp"
-  diff -q "$f" "$f.tmp" >/dev/null || mv "$f.tmp" "$f"
-  rm -f "$f.tmp"
-}
-export -f dos_to_unix
-
-find tests include source examples \
-  \( -name '*.cc' -o -name '*.h' -o -name '*.cu' -o -name '*.cuh' \) -print0 |
-  xargs -0 -n 1 -P 10 -I {} bash -c 'dos_to_unix "$@"' _ {}
+process "tests include source examples" ".*\.(cc|h|cu|cuh)" format_file
```
[the remark this answers is not in the document]
Comment (tamiko) [reply contrib/utilities/indent]: @bangerth find already has an implicit `^$` for all regex searches.

**Comment by tamiko** [comment]
@bangerth The new version also checks every `*.output` and `*.cmake` file for correct line ending and permissions. The output files that were changed are changed permission settings.

File: contrib/utilities/indent_common.sh [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+# In order to format .inst.in files, we need to replace \{ and \} by a
+# sentinel because clang-format happily strips away the backslash. Further,
+# there is a minor caveat: clang-format automatically renames a comment
+# after a closing bracket for a namespace to "} // namespace FooBar". Thus
+# use "namespace" as keyword:
+#
+
+format_inst()
+{
+    cp "$1" "$1.tmp${NONCE}"
```
[the remark this answers is not in the document]
Comment (tamiko) [reply contrib/utilities/indent_common.sh]: @bangerth This PR was intended as a refactoring + mild code cleanup. I am happy to rewrite the whole thing following your suggestions.

File: contrib/utilities/indent (hunk 1 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply contrib/utilities/indent]: I see, thanks!

**Comment by tamiko** [comment]
@bangerth Regarding the last file - well removing `\r` from a binary file is a terrible idea ;-)

**Comment by tamiko** [comment]
@bangerth I am happy to address all of your suggestions :-) Please re-review. In the meantime I will wrestle down my girlfriends mac and check whether the script works there as well.

**Comment by tamiko** [comment]
@bangerth New error messages:
```
***   No clang-format program found.
***
***   You can run the './contrib/utilities/download_clang_format'
***   script, or the './contrib/utilities/compile_clang_format' script 
***   to install a compatible binary into './contrib/utilities/programs'.
```
```
***   This indent script requires clang-format version 6.0,
***   but version 4.0  was found instead.
***
***   You can run the './contrib/utilities/download_clang_format'
***   script, or the './contrib/utilities/compile_clang_format' script 
***   to install a compatible binary into './contrib/utilities/programs'.
```

**Comment by bangerth** [comment]
Wrestling things away from girl friends is generally not recommended. We all need to have a life outside work :-)

**Comment by tamiko** [comment]
This is now also Mac OSX secure ;-)

**Comment by bangerth** [comment]
I suspect that running the various regression testers is not going to uncover anything useful here, so I'll merge once the basic CI test is through.

**Comment by tamiko** [comment]
@bangerth Please wait a minute.

**Comment by tamiko** [comment]
@bangerth I sneaked in one last change, https://github.com/dealii/dealii/pull/6762/commits/c35016a27cd68870a8728ea1829c17fc55304aa9 . Otherwise sed on Mac fails on its nose.

File: contrib/utilities/indent_common.sh [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+# In order to format .cc and .h files we have to make sure that we override
+# the source/header file only if the actual contents changed.
+# Unfortunately, clang-format isn't exactly helpful there. Thus, use a
+# temporary file and diff as a workaround.
+#
+
+format_file()
+{
+  file="${1}"
+  tmpfile="$(mktemp "${TMPDIR}/$(basename "$1").tmp.XXXXXXXX")"
```
Comment (bangerth) [inline contrib/utilities/indent_common.sh]: Thanks for this little change!

**Pull request merged**